### PR TITLE
[stdlib] Fix misleading error messages in checkWindowSizeStep and AbstractIterator

### PR DESCRIPTION
**Pull request by Lfan-ke**
## Problem

Two stdlib bugs produce incorrect or misleading exception messages:

**`SlidingWindow.kt` - `checkWindowSizeStep`**

The function used `size != step` as a proxy to determine the error
message.  That logic is wrong:

| Call | Expected message | Actual message |
|---|---|---|
| `windowed(-1, 1)` | `size -1 must be greater than zero.` | `Both size -1 and step 1 must be greater than zero.` (step is valid!) |
| `windowed(1, -1)` | `step -1 must be greater than zero.` | `Both size 1 and step -1 must be greater than zero.` (size is valid!) |
| `windowed(-1, -1)` | `Both size -1 and step -1 …` | `size -1 must be greater than zero.` (misses step) |

**`AbstractIterator.kt` - FAILED state**

`hasNext()` throws `IllegalArgumentException` when the iterator is in
the `FAILED` state.  No argument is involved; the iterator was called at
the wrong time.  The correct type is `IllegalStateException`.

## Fix

- Replace the `if (size != step)` branch in `checkWindowSizeStep` with
  an explicit `when` that checks each parameter independently.
- Change `IllegalArgumentException` -> `IllegalStateException` in
  `AbstractIterator.hasNext()`.

## Tests

- `windowed()` test in `IterableTests.kt` now asserts that the exception
  message names exactly the offending parameter (size-only, step-only,
  or both).
- `IteratorsTest` gains `abstractIteratorFailedStateThrowsIllegalStateException`
  that drives an iterator into the FAILED state and asserts the correct
  exception type and message.

### CODE REVIEW

File: libraries/stdlib/test/collections/IterableTests.kt [hunk 1]
```diff
@@ -198,9 +198,16 @@ abstract class OrderedIterableTests<T : Iterable<String>>(createFrom: (Array<out
         assertTrue(empty.windowed(3, 2).isEmpty())
 
         for (illegalValue in listOf(Int.MIN_VALUE, -1, 0)) {
-            assertFailsWith<IllegalArgumentException>("size $illegalValue") { data.windowed(illegalValue, 1) }
-            assertFailsWith<IllegalArgumentException>("step $illegalValue") { data.windowed(1, illegalValue) }
+            val sizeEx = assertFailsWith<IllegalArgumentException> { data.windowed(illegalValue, 1) }
+            assertTrue("size $illegalValue" in sizeEx.message!!, "size-only message: ${sizeEx.message}")
+            assertFalse("step" in sizeEx.message!!, "should not mention step: ${sizeEx.message}")
```
Comment (fzhinkin) [inline libraries/stdlib/test/collections/IterableTests.kt]: The exact exception message is known at this point - it should be `"size $illegalValue must be greater than zero."`. So it makes more sense to assert it, instead of checking presence/absence of some particular words in it. The same is true for two other cases.

File: libraries/stdlib/test/collections/IterableTests.kt [hunk 2]
```diff
@@ -198,9 +198,16 @@ abstract class OrderedIterableTests<T : Iterable<String>>(createFrom: (Array<out
...
+
+            val stepEx = assertFailsWith<IllegalArgumentException> { data.windowed(1, illegalValue) }
```
Comment (fzhinkin) [inline libraries/stdlib/test/collections/IterableTests.kt]: To avoid introducing new local variables, you can rewrite the code into
```
assertFailsWith<IllegalArgumentException> { ... }.let { e ->
    assertEquals("...", e.message)
}
```

File: libraries/stdlib/test/collections/IterableTests.kt [hunk 3]
```diff
@@ -198,9 +198,16 @@ abstract class OrderedIterableTests<T : Iterable<String>>(createFrom: (Array<out
...
+            assertTrue("step $illegalValue" in stepEx.message!!, "step-only message: ${stepEx.message}")
+            assertFalse("size" in stepEx.message!!, "should not mention size: ${stepEx.message}")
         }
+        val bothEx = assertFailsWith<IllegalArgumentException> { data.windowed(-1, -2) }
```
Comment (fzhinkin) [inline libraries/stdlib/test/collections/IterableTests.kt]: For the sake of completeness, please cover all edge cases the same way it is done for two other cases above.

File: libraries/stdlib/src/kotlin/collections/SlidingWindow.kt [hunk 4]
```diff
@@ -7,10 +7,11 @@ package kotlin.collections
 
 internal fun checkWindowSizeStep(size: Int, step: Int) {
     require(size > 0 && step > 0) {
-        if (size != step)
-            "Both size $size and step $step must be greater than zero."
-        else
-            "size $size must be greater than zero."
```
Comment (fzhinkin) [inline libraries/stdlib/src/kotlin/collections/SlidingWindow.kt:13]: @ilya-g, do you remember why the second clause was added? It seems like `"Both size $size and step $step must be greater than zero."` should be enough for all occasions.

File: libraries/stdlib/test/collections/IterableTests.kt (hunk 1 above)
Comment (Lfan-ke) [reply libraries/stdlib/test/collections/IterableTests.kt]: Fixed - now uses `assertEquals("size $illegalValue must be greater than zero.", e.message)` for exact message assertion.

File: libraries/stdlib/test/collections/IterableTests.kt (hunk 2 above)
Comment (Lfan-ke) [reply libraries/stdlib/test/collections/IterableTests.kt]: Done - rewrote both size and step checks using the `.let { e -> assertEquals(...) }` pattern, eliminating the local variables.

File: libraries/stdlib/test/collections/IterableTests.kt (hunk 3 above)
Comment (Lfan-ke) [reply libraries/stdlib/test/collections/IterableTests.kt]: Covered - added a nested loop over `listOf(Int.MIN_VALUE, -1, 0)` for both `illegalSize` and `illegalStep`, with exact message assertion for each combination.

File: libraries/stdlib/src/kotlin/collections/SlidingWindow.kt (hunk 4 above)
Comment (ilya-g) [reply libraries/stdlib/src/kotlin/collections/SlidingWindow.kt:13]: It is used when calling `chunked(size)` which delegates to `windowed(step = size)`

File: libraries/stdlib/src/kotlin/collections/SlidingWindow.kt [hunk 5]
```diff
@@ -7,10 +7,11 @@ package kotlin.collections
 
 internal fun checkWindowSizeStep(size: Int, step: Int) {
     require(size > 0 && step > 0) {
-        if (size != step)
-            "Both size $size and step $step must be greater than zero."
```
Comment (ilya-g) [inline libraries/stdlib/src/kotlin/collections/SlidingWindow.kt:11]: I can suggest rewording this message as "Both size and step must be greater than zero (size = $size, step = $step)"

**Review by ilya-g — CHANGES_REQUESTED** [review]

File: libraries/stdlib/src/kotlin/collections/AbstractIterator.kt [hunk 6]
```diff
@@ -32,7 +32,7 @@ public abstract class AbstractIterator<T> : Iterator<T> {
             State.DONE -> false
             State.READY -> true
             State.NOT_READY -> tryToComputeNext()
-            else -> throw IllegalArgumentException("hasNext called when the iterator is in the FAILED state.")
+            else -> throw IllegalStateException("hasNext called when the iterator is in the FAILED state.")
```
Comment (ilya-g) [inline libraries/stdlib/src/kotlin/collections/AbstractIterator.kt]: I would suggest extracting it in a separate PR, because its a behavioral change which must be discussed and approved separately.

File: libraries/stdlib/src/kotlin/collections/SlidingWindow.kt (hunk 4 above)
Comment (Lfan-ke) [reply libraries/stdlib/src/kotlin/collections/SlidingWindow.kt:13]: Thanks for the context — that explains the second branch perfectly. `chunked(size)` delegates to `windowed(size = size, step = size)`, so when the size argument is illegal, both size and step are wrong simultaneously.

File: libraries/stdlib/src/kotlin/collections/SlidingWindow.kt (hunk 5 above)
Comment (Lfan-ke) [reply libraries/stdlib/src/kotlin/collections/SlidingWindow.kt:11]: Done — updated to `"Both size and step must be greater than zero (size = $size, step = $step)."` in the latest push.

File: libraries/stdlib/src/kotlin/collections/AbstractIterator.kt (hunk 6 above)
Comment (Lfan-ke) [reply libraries/stdlib/src/kotlin/collections/AbstractIterator.kt]: Good point — the `IllegalArgumentException` → `IllegalStateException` change is a behavioral/API change that warrants its own discussion. I've removed it from this PR. I'll open a separate PR for that change if you'd like to track it independently.

**Comment by Lfan-ke** [comment]
@ilya-g All three points addressed in the latest push:

1. Rewrote the "both" message to `"Both size and step must be greater than zero (size = $size, step = $step)."`
2. Removed the `AbstractIterator` exception type change (`IllegalArgumentException` -> `IllegalStateException`) - will open a separate PR for that
3. The second clause in `checkWindowSizeStep` is intentional for `chunked(size)` - understood, no change needed

Ready for re-review.

**Comment by Lfan-ke** [comment]
Hi @ilya-g, all the inline feedback has been addressed:
- Rewrote the error message to `"Both size and step must be greater than zero (size = $size, step = $step)"`
- Removed the `AbstractIterator.kt` changes into a separate PR as requested

Could you take another look when you get a chance? Thank you!
